### PR TITLE
test-ng: add file.py for simple file tests 

### PR DESCRIPTION
**Pull request by yeoldegrove**
**What this PR does / why we need it**:

This PR adds a `file()` fixture that can check various file metadata:

##### File Type Checks

- `exists(path)` -> bool: Check if path exists (any type)
- `is_regular_file(path)` -> bool: Check if path is a regular file (not symlinks)
- `is_dir(path)` -> bool: Check if path is a directory
- `is_symlink(path, target=None, error_on_broken_symlink=True)` -> bool: Check if path is a symlink, optionally validating target

##### Permission Check

- `is_executable(path)` -> bool: Check if file is executable
- `is_readable(path)` -> bool: Check if file is readable
- `is_writable(path)` -> bool: Check if file is writable

##### Permission Metadata

- `has_mode(path, mode: str)` -> bool: Check if file has specific permissions (e.g., "0755")
- `get_mode(path)` -> str: Get file permissions as octal string (e.g., "0755")

##### Ownership Checks

- `is_owned_by_user(path, user: str)` -> bool: Check if file is owned by specific user
- `is_owned_by_group(path, group: str)` -> bool: Check if file is owned by specific group
- `is_owned_by(path, user: str, group: str)` -> bool: Check if file is owned by specific user and group

##### Ownership Metadata

- `get_owner(path)` -> tuple[str, str]: Get (user, group) ownership
- `get_user(path)` -> str: Get file owner username
- `get_group(path)` -> str: Get file owner group name

##### Other

- `get_size(path)` -> int: Get file size in bytes

##### Tests for plugins

- To discuss if needed or overkill:

Same as in #3881...
- **Plugin Testing**: Added 61 tests covering all methods, formats, edge cases, and error conditions for the `file.py`  plugin.
  - Idea: `dev_tests.yml` workflow or a developer runs `make -f tests-ng/dev.makefile test-plugins` to run integration tests of plugins. `tests-ng/plugins/tests/test_*.py` can contain tests that are run automatically after each push or locally.

**Which issue(s) this PR fixes**:
Fixes #3904

### CODE REVIEW

File: tests-ng/plugins/file.py [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+
+        Args:
+            path: File path.
+
+        Returns:
+            bool: True if the path exists, False otherwise.
+        """
+        return Path(path).exists()
+
+    def is_file(self, path: str | Path) -> bool:
```
Comment (NotTheEvilOne) [inline tests-ng/plugins/file.py]: The method name is misleading. Looking at `pathlib.Path.is_file()` as a similar one would imply that this function returns true as long as a path exists and is not a directory. As the description and code indicates this is true for regular files instead only. Maybe you find a better name for the method?

File: tests-ng/plugins/file.py [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+
+        Args:
+            path: File path.
+
+        Returns:
+            bool: True if the path is a directory, False otherwise.
+        """
+        p = Path(path)
+        try:
+            return stat.S_ISDIR(p.lstat().st_mode)
```
Comment (NotTheEvilOne) [inline tests-ng/plugins/file.py]: `Path.is_dir(follow_symlinks=False)` should return the same result. Please note that `stat()` is usually more expensive (time consuming / memory wise).

File: tests-ng/plugins/file.py [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+
+        Raises:
+            FileNotFoundError: If error_on_broken_symlink is True and the symlink
+                target does not exist.
+        """
+        p = Path(path)
+        if not p.is_symlink():
+            return False
+
+        if target is not None:
```
Comment (NotTheEvilOne) [inline tests-ng/plugins/file.py]:
```
if target is None:
  return True
```

Would make the code more readable here :)

File: tests-ng/plugins/file.py [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+            symlink_target = p.readlink()
+            target_path = Path(target)
```
Comment (NotTheEvilOne) [inline tests-ng/plugins/file.py]: I think `pathlib.Path.samefile()` should give you the same logic here. Would be cool if one line could replace the rest of the method :)

File: tests-ng/plugins/file.py [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+        """
+        stat_info = Path(
+            path
+        ).stat()  # Raises FileNotFoundError or PermissionError naturally
+        mode = stat_info.st_mode
+        current_uid = os.getuid()
+        current_gids = {os.getgid()} | set(os.getgroups())
+
+        # Map permission types to (owner, group, other) bit masks
+        permission_bits = {
```
Comment (NotTheEvilOne) [inline tests-ng/plugins/file.py]: This could be constants (e.g. top of file) removing the need to map memory for each call to the method.

Not sure if something like a contant like PERMISSIONS_READ and a generic approach in this method to what is actually checked would be more suitable than a string like "read", "write" or "execute". Could be octets given as inputs like 0x700 etc.

File: tests-ng/plugins/file.py [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+            path: File path.
+
+        Returns:
+            bool: True if the file is executable, False otherwise.
+
+        Raises:
+            FileNotFoundError: If the path does not exist.
+            PermissionError: If permission to access the path is denied.
+        """
+        return self._check_permission(path, "execute")
```
Comment (NotTheEvilOne) [inline tests-ng/plugins/file.py]: `os.access` with `os.X_OK` should be used here as for portability and other reasons I guess more than just file permissions are relevant to return a correct result if a path is executable or not.

File: tests-ng/plugins/file.py [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+            path: File path.
+
+        Returns:
+            bool: True if the file is writable, False otherwise.
+
+        Raises:
+            FileNotFoundError: If the path does not exist.
+            PermissionError: If permission to access the path is denied.
+        """
+        return self._check_permission(path, "write")
```
Comment (NotTheEvilOne) [inline tests-ng/plugins/file.py]: Same as for `is_executable()`.

File: tests-ng/plugins/file.py [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+            path: File path.
+
+        Returns:
+            bool: True if the file is readable, False otherwise.
+
+        Raises:
+            FileNotFoundError: If the path does not exist.
+            PermissionError: If permission to access the path is denied.
+        """
+        return self._check_permission(path, "read")
```
Comment (NotTheEvilOne) [inline tests-ng/plugins/file.py]: Same as for `is_executable()`.

File: tests-ng/plugins/file.py [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+
+        Returns:
+            Tuple[str, str]: A tuple of (username, groupname).
+
+        Raises:
+            FileNotFoundError: If the path does not exist.
+            PermissionError: If permission to access the path is denied.
+        """
+        stat_info = Path(path).stat()
+        user = pwd.getpwuid(stat_info.st_uid).pw_name
```
Comment (NotTheEvilOne) [inline tests-ng/plugins/file.py]: Should call `self.get_user()` / `self.get_group()` for easier maintainability. This way only these methods need to be changed if needed.

File: tests-ng/plugins/file.py [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+            str: Group name of the file owner.
+
+        Raises:
+            FileNotFoundError: If the path does not exist.
+            PermissionError: If permission to access the path is denied.
+        """
+        stat_info = Path(path).stat()
+        return grp.getgrgid(stat_info.st_gid).gr_name
+
+    def has_user(self, path: str | Path, user: str) -> bool:
```
Comment (NotTheEvilOne) [inline tests-ng/plugins/file.py]: This should maybe be called `is_owned_by_user()`.

File: tests-ng/plugins/file.py [hunk 11]
```diff
@@ -0,0 +1,368 @@
...
+        Returns:
+            bool: True if the file has the specified user, False otherwise.
+
+        Raises:
+            FileNotFoundError: If the path does not exist.
+            PermissionError: If permission to access the path is denied.
+        """
+        return self.get_user(path) == user
+
+    def has_group(self, path: str | Path, group: str) -> bool:
```
Comment (NotTheEvilOne) [inline tests-ng/plugins/file.py]: This should maybe be called `is_owned_by_group()`.

File: tests-ng/plugins/file.py [hunk 12]
```diff
@@ -0,0 +1,368 @@
...
+        Returns:
+            bool: True if the file has the specified group, False otherwise.
+
+        Raises:
+            FileNotFoundError: If the path does not exist.
+            PermissionError: If permission to access the path is denied.
+        """
+        return self.get_group(path) == group
+
+    def has_owner(self, path: str | Path, user: str, group: str) -> bool:
```
Comment (NotTheEvilOne) [inline tests-ng/plugins/file.py]: This should maybe be called `is_owned_by()` accepting user and group given.

File: tests-ng/plugins/file.py (hunk 1 above)
Comment (yeoldegrove) [reply tests-ng/plugins/file.py]: Thanks for noticing this inconsistency. `is_regular_file()` comes to mind. Will change it.

File: tests-ng/plugins/file.py (hunk 2 above)
Comment (yeoldegrove) [reply tests-ng/plugins/file.py]: Nice suggestion, will change it.

File: tests-ng/plugins/file.py (hunk 3 above)
Comment (yeoldegrove) [reply tests-ng/plugins/file.py]: Will move the explicit "is None" a bit up.

Comment (yeoldegrove) [reply tests-ng/plugins/file.py]: Reworked 'is_symlink()' for better readability instead. Even `if target is None` we need to check for broken symlinks.

File: tests-ng/plugins/file.py (hunk 4 above)
Comment (yeoldegrove) [reply tests-ng/plugins/file.py]: We need some logic to find broken symlinks and non-matching targets and `pathlib.Path.samefile()` won't cover this automatically.
Nevertheless I reworked `is_symlink()` for better readability.

File: tests-ng/plugins/file.py (hunk 5 above)
Comment (yeoldegrove) [reply tests-ng/plugins/file.py]: Removed `_check_permissions()` completely.

File: tests-ng/plugins/file.py (hunk 6 above)
Comment (yeoldegrove) [reply tests-ng/plugins/file.py]: Great hint. This made the whole `_check_permissions()` function obsolete.

File: tests-ng/plugins/file.py (hunk 7 above)
Comment (yeoldegrove) [reply tests-ng/plugins/file.py]: :+1:

File: tests-ng/plugins/file.py (hunk 8 above)
Comment (yeoldegrove) [reply tests-ng/plugins/file.py]: :+1:

File: tests-ng/plugins/file.py (hunk 9 above)
Comment (yeoldegrove) [reply tests-ng/plugins/file.py]: :+1:

File: tests-ng/plugins/file.py (hunk 10 above)
Comment (yeoldegrove) [reply tests-ng/plugins/file.py]: Yes, the renaming helps understanding what the function does. It is a bit long but this can be sacrificed for better clarity, I guess.

File: tests-ng/plugins/file.py (hunk 11 above)
Comment (yeoldegrove) [reply tests-ng/plugins/file.py]: Yes, the renaming helps understanding what the function does. It is a bit long but this can be sacrificed for better clarity, I guess.

File: tests-ng/plugins/file.py (hunk 12 above)
Comment (yeoldegrove) [reply tests-ng/plugins/file.py]: Yes, the renaming helps understanding what the function does. It is a bit long but this can be sacrificed for better clarity, I guess.

**Comment by yeoldegrove** [comment]
@NotTheEvilOne Adressed all your comments, please re-review.

File: .github/workflows/dev_test-ng.yml
Comment (NotTheEvilOne) [inline .github/workflows/dev_test-ng.yml:1]: Would it make sense to rename this workflow to `test_*.yml` as reuse in other workflows would have consistent naming this way.

Comment (yeoldegrove) [reply .github/workflows/dev_test-ng.yml:1]: @NotTheEvilOne I wanted to destuinguish it from `test_*.yml` as this is not about running tests on any of our artifacts or reusing it in other workflows but suits a development purpose.

Maybe re-naming it `dev_tests.yml` and combining it with `lint_tests-ng.yml` would be less ugly. Those two serve a similar purpose.
